### ISSUE TITLE
PP-ct permissions

### ISSUE THREAD

**Issue by eileenmcnaughton** [eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor#108]
The current status with permissions is that the preapprove api has pretty broad permissions ('make online contributions') which is fine since it doesn't actually allow credit card testing. However, it turns out it DOES require 'access Ajax api' because it's an api call - what is the best recommendation here. We DO need 'something'

**Comment by eileenmcnaughton** [comment]
@colemanw @mlutfy @totten looks like the paypal checkout page on civicrm.org https://civicrm.org/civicrm/contribute/transact?reset=1&id=107 doesn't work logged out & I believe it's permissions per above - what are your thoughts it seems we either have to give anon 'access Ajax api' or come with another way to call the api without requiring that....

**Comment by colemanw** [comment]
If the payment processor requires the api then yes anonymous users will need permission to access it.
I don't think this is a big security worry as the api will still check permissions.

**Comment by totten** [comment]
@eileenmcnaughton I'm coming into this conversation a little cold. For a couple sites and/or near-term time-frame, granting `access ajax api` does seem ok. I tried grepping a bit to try to find APIs that might inadvertently get exposed. (`contact.getquick` came up. The code looks ACL-aware, though I only skimmed.)

If this is being deployed more broadly, it is a bit of config work. Consider do something programmatic to nudge the site to the needful, e.g. one of:

* Implement `hook_civicrm_check`. Do a self-test (eg `curl` or `file_get_contents()`) to see if the end-point is effectively accessible. Show a warning if it's not.
* Implement `hook_civircm_alterPermissions` and grant `access ajax api` programmatically. (To tighten, maybe inspect `$_REQUEST` to see if we're hitting the AJAX end point with a specific/target API -- only grant `access ajax api` if the specific API is supported.)
* Implement `hook_civicrm_alterMenu` and modify (maybe even wrap) the AJAX end-point (to the same general end/effect as above).
